src/v/cloud_topics/reconciler/reconciliation_consumer.cc

```diff
@@ -16,47 +16,33 @@
 
 namespace cloud_topics::reconciler {
 
-reconciliation_consumer::reconciliation_consumer(
-  l1::object_builder* builder, model::topic_id_partition tidp)
-  : _builder(builder)
-  , _tidp(tidp)
-  , _metadata{
-      .base_offset = kafka::offset::min(),
-      .last_offset = kafka::offset::min(),
-      .last_timestamp = model::timestamp::min(),
-      .batch_count = 0} {}
-
-ss::future<ss::stop_iteration>
-reconciliation_consumer::operator()(model::record_batch batch) {
-    if (_metadata.base_offset == kafka::offset::min()) {
-        _metadata.base_offset = model::offset_cast(batch.base_offset());
-        co_await _builder->start_partition(_tidp);
-    }
-
-    // NOTE: Only data batches here. It's safe to use timestamps without
-    //       checking the batch type.
-    _metadata.last_timestamp = std::max(
-      batch.header().max_timestamp, _metadata.last_timestamp);
-    _metadata.last_offset = model::offset_cast(batch.last_offset());
-
-    if (!_metadata.terms.contains(batch.term())) {
-        _metadata.terms.insert(
-          std::make_pair(
-            batch.term(), model::offset_cast(batch.base_offset())));
-    }
-
-    ++_metadata.batch_count;
-
-    co_await _builder->add_batch(std::move(batch));
-
-    co_return ss::stop_iteration::no;
-}
-
-std::optional<consumer_metadata> reconciliation_consumer::end_of_stream() {
-    if (_metadata.base_offset != kafka::offset::min()) {
-        return _metadata;
+ss::future<std::optional<consumer_metadata>> build_from_reader(
+  model::topic_id_partition tidp,
+  model::record_batch_reader reader,
+  l1::object_builder* builder) {
+    auto gen = std::move(reader).slice_generator(model::no_timeout);
+    co_await builder->start_partition(tidp);
```

Comment:
> This will allow us to add metrics to track the
difference between L0 read time and the staging file write time.

why is the slice generator necessary to achieve this? if you would have measured the costs of co_awaiting from the generator then you'd have still measured the correct aggregate time you'd just have a bunch of near zero measurements. now we have a bit of a reader contract we didn't have before, namely that the reader shouldn't be using its own internal slice-like caching less the new measurements here won't be measuring what was intended.

Reply:
What I mean is that it's not accurate to claim that obtaining a new slice always performs I/O (as mentioned in the commit message). For example, a reader could maintain it's own slice or cache (like the log_reader).

> Yeah fair I guess I was over optimizing here to not take a bunch of near 0 measurements.

However, if we are still just measuring invocations of the obtain-slice and adding up the results, the effect should be similar and the measurement would encompass the near-zero invocations as well as the invocations that do actual I/O, without taking a measurement for every batch, as you mention.